File: 프로그래머스/unrated/142085. 디펜스 게임/디펜스 게임.py

def check(enemy , n, k):
    if k >= len(enemy):
        return True 

    enemy.sort(reverse=True)
    
    for i in range(k):
        enemy[i] = 0
    
    if sum(enemy) <= n:
        return True
    else:
        return False
        
def solution(n, k, enemy):
    left = 0
    right = len(enemy) + 1

    while( left + 1 < right ):
        mid = (left + right) // 2
        if check(enemy[:mid], n, k):
            left = mid 
        else:
            right = mid
    
    return left

File: 6.programmers_auto_upload/unrated/142085. 디펜스 게임/디펜스 게임.java
import java.util.Arrays;

class Solution {
    
    public boolean check(int[] enemy, int n, int k){
        if ( k >= enemy.length ){
            return true;
        }
        for(int i = 0; i < k; i++){
            enemy[i] = 0;
        }
        int sum = 0;
        for (int e : enemy){
            sum += e;
        }
        return sum <= n;
    }
    
    public int solution(int n, int k, int[] enemy) {
        int left = 0;
        int right = enemy.length+1;

        while( left + 1 < right){
            int mid = (left + right) / 2;
            int[] current_rocks = Arrays.copyOfRange(enemy, 0, mid);
            if (check(current_rocks, n, k)) {
                left = mid;
            }else{
                right = mid;
            }
        }
    }
    
    
}